File: packages/runner/src/run-case.ts
import {
  applyOverrides,
  DEFAULT_EVIDENCE_PLAN,
  DEFAULT_VERDICT_POLICY,
  evaluateVerdict,
  type Baseline,
  type EvidencePlan,
  type Expectation,
  type Observation,
  type TestCase,
  type Verdict,
  type VerdictPolicy,
  type VerdictResult,
} from "@agentic-test-hub/core";

import { deriveActionParams } from "./derive-params.ts";
import {
  COLLECTED_BY_OPERATION,
  observeBrowser,
  observeFromResult,
  type ObserveOptions,
} from "./evidence.ts";
import { checkExpectation } from "./expectation.ts";
import { prepareStates, type PreparationReport } from "./state.ts";
import type { AssertionOutcome } from "./assert.ts";
import type { BrowserSession } from "./executor/browser.ts";
import type { ExecutorRegistry } from "./executor/registry.ts";
import type { ExecutionContext, ExecutionResult } from "./executor/types.ts";

/** One expectation, and what became of it. */
export interface ExpectationOutcome {
  readonly expectation: Expectation;
  readonly outcome: AssertionOutcome;
}

/** Options shared by `runCase` and `runStep`/`runScenario`. */
export interface RunOptions {
  /**
   * A browser session already open for this run, for the two channels that
   * exist only while a page is open. Never opened or closed here — that is
   * the caller's lifecycle, same convention as `BrowserExecutor`'s own
   * borrowed-session mode.
   */
  readonly browserSession?: BrowserSession;
  /** Overrides the manifest's own policy, which overrides `DEFAULT_VERDICT_POLICY`. */
  readonly policy?: VerdictPolicy;
  /** Noise filtering passed through to evidence collection. */
  readonly observe?: ObserveOptions;
}

/** Why a case could not even be attempted. */
export interface Blocked {
  readonly reason: string;
  /** Override paths that have no mechanical path to the executed request. */
  readonly paths: readonly string[];
}

/** What running one case established. */
export interface CaseRunResult {
  readonly caseId: string;
  readonly preparation: PreparationReport;
  /** Set when the case's overrides could not be mechanically derived; nothing past this point ran. */
  readonly blocked?: Blocked;
  readonly action?: ExecutionResult;
  readonly expectations: readonly ExpectationOutcome[];
  readonly observations: readonly Observation[];
  readonly evidence: VerdictResult;
  readonly verdict: Verdict;
}

const VERDICT_RANK: Record<Verdict, number> = { pass: 0, inconclusive: 1, fail: 2 };

/** The worst of several verdicts, `pass` being the best case. */
export function worstVerdict(verdicts: readonly Verdict[]): Verdict {
  return verdicts.reduce<Verdict>(
    (worst, verdict) => (VERDICT_RANK[verdict] > VERDICT_RANK[worst] ? verdict : worst),
    "pass",
  );
}

function verdictOfOutcome(outcome: AssertionOutcome): Verdict {
  if (outcome.verdict === "violated") return "fail";
  if (outcome.verdict === "needs_judgement") return "inconclusive";
  return "pass";
}

/** Composes expectation outcomes and, optionally, an evidence verdict into one overall verdict. */
export function composeVerdict(
  expectations: readonly ExpectationOutcome[],
  evidence?: VerdictResult,
): Verdict {
  const verdicts = expectations.map((entry) => verdictOfOutcome(entry.outcome));
  if (evidence !== undefined) verdicts.push(evidence.verdict);
  return worstVerdict(verdicts);
}

/**
 * Collects the evidence a plan calls for, given what is actually available.
 *
 * `db_records`/`app_log`/`db_log` are gathered by running their declared
 * collector operation and scanning its output for problem words (see
 * `observeFromResult`); `browser_console`/`browser_network` come from a live
 * session instead, since they do not exist once it closes. `screenshot` is
 * never gathered here — deciding whether a page *looks* right is a judgement
 * this module does not make on its own (see `evidence.ts`'s own docs on
 * this). Deliberately no attempt is made to synthesise `db_records`'
 * `intendedChanges`/`residualChanges` beyond what `observeFromResult`
 * already reports (nothing today implements real before/after row diffing),
 * so a policy that binds `db_records` to `expected_change`/
 * `no_residual_change` will not pass on this alone — an honest limitation,
 * not a bug.
 */
export async function collectEvidence(
  plan: EvidencePlan,
  registry: ExecutorRegistry,
  context: ExecutionContext,
  options: RunOptions,
): Promise<Observation[]> {
  const observations: Observation[] = [];
  if (options.browserSession !== undefined) {
    observations.push(...observeBrowser(options.browserSession, options.observe));
  }
  for (const source of COLLECTED_BY_OPERATION) {
    if (!plan.sources.includes(source)) continue;
    const call = context.manifest.evidence[source];
    if (call === undefined) continue;
    const result = await registry.run(call.operation, call.params, context);
    observations.push(observeFromResult(source, result, options.observe));
  }
  return observations;
}

/**
 * Runs one test case: resolves it against its baseline, prepares its
 * preconditions, runs its action, judges every expectation, collects
 * evidence and reaches an overall verdict.
 *
 * This is the one function both a generated Playwright test and any future
 * direct-execution path (a CLI, a hub "run now" button) call — generated
 * code's own job is only to load a manifest, build a registry and context,
 * and call this.
 *
 * @param testCase - The case to run.
 * @param baseline - Its baseline, already looked up.
 * @param registry - Executors available to run its operations with.
 * @param context - Manifest, scopes and cancellation.
 * @param options - A browser session to read, a policy override, noise
 *   filtering.
 * @returns What preparing, running and judging the case established.
 */
export async function runCase(
  testCase: TestCase,
  baseline: Baseline,
  registry: ExecutorRegistry,
  context: ExecutionContext,
  options: RunOptions = {},
): Promise<CaseRunResult> {
  const policy = options.policy ?? context.manifest.policy ?? DEFAULT_VERDICT_POLICY;
  const resolved = applyOverrides(baseline, testCase.overrides);

  const preparation = await prepareStates(resolved.preconditions, registry, context);
  if (!preparation.ready) {
    return {
      caseId: testCase.id,
      preparation,
      expectations: [],
      observations: [],
      evidence: evaluateVerdict(policy, testCase.polarity, []),
      verdict: "inconclusive",
    };
  }

  if (resolved.action === undefined) {
    throw new Error(`case ${testCase.id}'s baseline declares no action to run`);
  }
  const operation = context.manifest.operations[resolved.action.operation];
  if (!operation) {
    throw new Error(
      `case ${testCase.id}'s baseline names operation ${resolved.action.operation}, which plugin "${context.manifest.name}" does not declare`,
    );
  }

  const derivation = deriveActionParams(
    testCase.overrides,
    resolved,
    resolved.action.operation,
    operation,
  );
  if (!derivation.ok) {
    return {
      caseId: testCase.id,
      preparation,
      blocked: {
        reason: "one or more overrides have no mechanical path to the operation they resolve to",
        paths: derivation.unexpressible.map((entry) => entry.path),
      },
      expectations: [],
      observations: [],
      evidence: evaluateVerdict(policy, testCase.polarity, []),
      verdict: "inconclusive",
    };
  }

  const action = await registry.run(resolved.action.operation, derivation.params, context);

  const expectations: ExpectationOutcome[] = [];
  for (const expectation of testCase.expect) {
    const subject =
      expectation.kind === "operation_result"
        ? await registry.run(expectation.operation, expectation.params, context)
        : action;
    expectations.push({ expectation, outcome: checkExpectation(expectation, subject) });
  }

  const plan = testCase.evidence ?? DEFAULT_EVIDENCE_PLAN;
  const observations = await collectEvidence(plan, registry, context, options);
  const evidence = evaluateVerdict(
    policy,
    testCase.polarity,
    observations,
    testCase.evidenceWaivers,
  );

  return {
    caseId: testCase.id,
    preparation,
    action,
    expectations,
    observations,
    evidence,
    verdict: composeVerdict(expectations, evidence),
  };
}

File: python/agentic-test-hub-runner/src/agentic_test_hub_runner/run_case.py
"""
Running one resolved case: preparing its preconditions, running its action,
judging every expectation, collecting evidence and reaching a verdict.
Mirrors `packages/runner/src/run-case.ts`, with one deliberate difference:

**This module never resolves overrides.** `packages/runner`'s own
`runCase(testCase, baseline, ...)` calls `applyOverrides` itself, because
that logic — and the schema it operates on — is TypeScript-only by design
(see the plan this package was built from: "resolution/derivation stays
TS-only ... run once at generation time"). By the time a Python test calls
this module, `ath-generate-test` has already computed the resolved baseline
(`ResolvedBaseline`, inlined by the CLI's Python template as `RESOLVED`) and
the mechanically-derived action params (`deriveActionParams`, inlined as
`ACTION_PARAMS`) — this function's own job is only to *run* that already-
resolved data and judge the result, exactly like the TypeScript version does
after its own `applyOverrides` call returns.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from .assertion import AssertionOutcome
from .evidence import (
    COLLECTED_BY_OPERATION,
    BrowserSession,
    ObserveOptions,
    observe_browser,
    observe_from_result,
)
from .expectation import check_expectation
from .policy import DEFAULT_VERDICT_POLICY, VerdictPolicy
from .registry import ExecutorRegistry
from .state import PreparationReport, prepare_states
from .types import ExecutionContext, ExecutionResult
from .verdict import EvidenceWaiver, Observation, Verdict, VerdictResult, evaluate_verdict

_VERDICT_RANK = {"pass": 0, "inconclusive": 1, "fail": 2}


def worst_verdict(verdicts: list[Verdict]) -> Verdict:
    """The worst of several verdicts, `pass` being the best case."""
    worst: Verdict = "pass"
    for verdict in verdicts:
        if _VERDICT_RANK[verdict] > _VERDICT_RANK[worst]:
            worst = verdict
    return worst


def _verdict_of_outcome(outcome: AssertionOutcome) -> Verdict:
    if outcome.verdict == "violated":
        return "fail"
    if outcome.verdict == "needs_judgement":
        return "inconclusive"
    return "pass"


@dataclass(frozen=True)
class ExpectationOutcome:
    expectation: dict[str, Any]
    outcome: AssertionOutcome


def compose_verdict(
    expectations: list[ExpectationOutcome], evidence: VerdictResult | None = None
) -> Verdict:
    """Composes expectation outcomes and, optionally, an evidence verdict."""
    verdicts = [_verdict_of_outcome(entry.outcome) for entry in expectations]
    if evidence is not None:
        verdicts.append(evidence.verdict)
    return worst_verdict(verdicts)


@dataclass(frozen=True)
class RunOptions:
    """Options shared by `run_case` and `run_step`/`run_scenario`."""

    browser_session: BrowserSession | None = None
    policy: VerdictPolicy | None = None
    observe: ObserveOptions | None = None


@dataclass(frozen=True)
class CaseRunResult:
    """What running one case established."""

    case_id: str
    preparation: PreparationReport
    action: ExecutionResult | None
    expectations: list[ExpectationOutcome]
    observations: list[Observation]
    evidence: VerdictResult
    verdict: Verdict


def collect_evidence(
    plan: dict[str, Any], registry: ExecutorRegistry, context: ExecutionContext, options: RunOptions
) -> list[Observation]:
    """
    Collects the evidence a plan calls for.

    `db_records`/`app_log`/`db_log` are gathered by running their declared
    collector operation; `browser_console`/`browser_network` come from a
    live session (never opened or closed here); `screenshot` is never
    gathered here, same limitation as the TypeScript original.
    """
    observations: list[Observation] = []
    if options.browser_session is not None:
        observations.extend(observe_browser(options.browser_session, options.observe))
    for source in COLLECTED_BY_OPERATION:
        if source not in plan.get("sources", []):
            continue
        call = context.manifest.evidence.get(source)
        if call is None:
            continue
        result = registry.run(call["operation"], call.get("params", {}), context)
        observations.append(observe_from_result(source, result, options.observe))
    return observations


_DEFAULT_EVIDENCE_PLAN = {
    "sources": [
        "screenshot",
        "browser_console",
        "browser_network",
        "db_records",
        "app_log",
        "db_log",
    ],
    "timing": "after",
}


def run_case(
    test_case: dict[str, Any],
    resolved: dict[str, Any],
    action_params: dict[str, Any],
    registry: ExecutorRegistry,
    context: ExecutionContext,
    options: RunOptions | None = None,
) -> CaseRunResult:
    """
    Runs one already-resolved case: prepares its preconditions, runs its
    action with `action_params`, judges every expectation, collects
    evidence and reaches an overall verdict.

    @param test_case: The case's own data (`id`, `expect`, `polarity`,
        `evidence`, `evidenceWaivers`) — its `overrides`/`baseline` fields,
        if present, are not read here; they were already applied when
        `resolved`/`action_params` were computed at generation time.
    @param resolved: The baseline with overrides already applied
        (`preconditions`, `action`, `context`, `config`) — `ResolvedBaseline`
        inlined by the CLI.
    @param action_params: The mechanically-derived params for `resolved`'s
        action, from `deriveActionParams` at generation time.
    """
    options = options or RunOptions()
    policy = options.policy or context.manifest.policy or DEFAULT_VERDICT_POLICY

    preparation = prepare_states(resolved.get("preconditions", []), registry, context)
    if not preparation.ready:
        return CaseRunResult(
            case_id=test_case["id"],
            preparation=preparation,
            action=None,
            expectations=[],
            observations=[],
            evidence=evaluate_verdict(policy, test_case.get("polarity", "nominal"), []),
            verdict="inconclusive",
        )

    action_ref = resolved.get("action")
    if action_ref is None:
        raise ValueError(f"case {test_case['id']}'s baseline declares no action to run")

    action = registry.run(action_ref["operation"], action_params, context)

    expectations: list[ExpectationOutcome] = []
    for expectation in test_case["expect"]:
        if expectation["kind"] == "operation_result":
            subject = registry.run(expectation["operation"], expectation.get("params", {}), context)
        else:
            subject = action
        expectations.append(
            ExpectationOutcome(expectation, check_expectation(expectation, subject))
        )

    plan = test_case.get("evidence") or _DEFAULT_EVIDENCE_PLAN
    observations = collect_evidence(plan, registry, context, options)
    waivers = [
        EvidenceWaiver(source=waiver["source"], reason=waiver["reason"])
        for waiver in test_case.get("evidenceWaivers", [])
    ]
    evidence = evaluate_verdict(policy, test_case.get("polarity", "nominal"), observations, waivers)

    return CaseRunResult(
        case_id=test_case["id"],
        preparation=preparation,
        action=action,
        expectations=expectations,
        observations=observations,
        evidence=evidence,
        verdict=compose_verdict(expectations, evidence),
    )
